File: cpp/basix/sobolev-spaces.cpp
// Copyright (c) 2022 Matthew Scroggs
// FEniCS Project
// SPDX-License-Identifier:    MIT

#include "sobolev-spaces.h"

basix::sobolev::space basix::sobolev::space_intersection(space space1,
                                                         space space2)
{
  if (space1 == space2)
    return space1;

  if (space1 == space::H1)
  {
    if (space2 == space::H2 or space2 == space::H3 or space2 == space::HInf)
      return space1;
  }
  else if (space1 == space::H2)
  {
    if (space2 == space::H1)
      return space2;
    else if (space2 == space::H3 or space2 == space::HInf)
      return space1;
  }
  else if (space1 == space::H3)
  {
    if (space2 == space::H1 or space2 == space::H2)
      return space2;
    else if (space2 == space::H3 or space2 == space::HInf)
      return space1;
  }
  else if (space1 == space::HInf)
  {
    if (space2 == space::H1 or space2 == space::H2 or space2 == space::H3)
      return space2;
  }

  return space::L2;
}

File: python/basix/sobolev_spaces.py
# Copyright (C) 2023-2024 Matthew Scroggs and Garth N. Wells
#
# This file is part of Basix (https://www.fenicsproject.org)
#
# SPDX-License-Identifier:    MIT
"""Functions for handling Sobolev spaces."""

from basix._basixcpp import SobolevSpace
from basix._basixcpp import sobolev_space_intersection as _ssi

__all__ = ["intersection"]


def intersection(spaces: list[SobolevSpace]) -> SobolevSpace:
    """Compute the intersection of a list of Sobolev spaces.

    Args:
        spaces: List of Sobolev spaces.

    Returns:
        Intersection of the Sobolev spaces.
    """
    space = spaces[0]
    for s in spaces[1:]:
        space = _ssi(space, s)
    return SobolevSpace[space.name]
